tower-http/src/compression/compression_filter.rs

```diff
@@ -0,0 +1,26 @@
+use crate::compression_utils::default_compression_filter_predicate;
+
+/// A filter which any response Parts needs to pass to be compressed
+pub trait CompressionFilter: Copy {
+    /// Predicate which takes response parts and returns true if the response should be compressed
+    fn should_compress(&self, parts: &http::response::Parts) -> bool;
+}
+
+impl<F> CompressionFilter for F where F: Fn(&http::response::Parts) -> bool + Copy {
+    fn should_compress(&self, parts: &http::response::Parts) -> bool {
+        (self)(parts)
+    }
+}
+
+/// Default compression filter that proxies `default_compression_filter_predicate` which looks at
+/// headers to determine whether compression is suitable
+#[derive(Default, Copy, Clone, Debug)]
+pub struct DefaultCompressionFilter {
+
+}
+
+impl CompressionFilter for DefaultCompressionFilter {
+    fn should_compress(&self, parts: &http::response::Parts) -> bool {
+        default_compression_filter_predicate(&parts.headers)
```

Comment:
What do you think about inlining `default_compression_filter_predicate` here? Might be slightly simpler since we don't have who places to look.

Reply:
It's a bit problematic since compression looks at the response while decompression looks at the request, since we're using `http::Response<B>` instead of headers they can't be easily merged

Reply:
Wait a minute. Why does decompression need to determined by a predicate? Shouldn't we always decompress things if they're compressed?

Reply:
The client advertising that it supports compression should always be fine since its up to the server if it actually wants to compress a resource. I think those checks are basically just for additional safety but could perhaps be removed. I looked at [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests) and [the spec](https://httpwg.org/specs/rfc7233.html#header.range) and they don't say `accept-encoding` cannot be combined with `range`.

Reply:
I think we should wait with touching decompression in this PR at least. To keep the scope smaller.